### PR TITLE
[TS migration] Migrate 'WorkspaceTravel' page to TypeScript

### PR DESCRIPTION
**Pull request by pac-guerreiro**
### Fixed Issues

$https://github.com/Expensify/App/issues/25192
PROPOSAL: N/A

### Tests

1. Go to `Settings > Workspaces`
2. Create a new workspace
3. On the workspace settings, click on `Travel`
4. Confirm that you're shown a section which gives you the option to `Connect bank account`
5. Click on `Connect bank account` and follow the steps
6. Confirm that you're shown a section which gives you the option to `Issue and manage cards`, `Book travel with concierge` and `Learn more`.

### Offline tests

Same as tests

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot_1706643937
Screenshot_1706643948

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot_1706642577
Screenshot_1706642591

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - Expensify - App - 2024-01-30 at 19 44 14
Simulator Screenshot - Expensify - App - 2024-01-30 at 19 44 29

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - Expensify - Safari - 2024-01-30 at 19 19 37
Simulator Screenshot - Expensify - Safari - 2024-01-30 at 19 19 53

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-01-30 at 19 17 28
Screenshot 2024-01-30 at 19 17 50

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2024-01-30 at 19 33 14
Screenshot 2024-01-30 at 19 33 26

</details>

### CODE REVIEW

File: src/pages/workspace/travel/WorkspaceTravelNoVBAView.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import {View} from 'react-native';
+import ConnectBankAccountButton from '@components/ConnectBankAccountButton';
+import * as Illustrations from '@components/Icon/Illustrations';
+import Section from '@components/Section';
+import Text from '@components/Text';
+import useLocalize from '@hooks/useLocalize';
+import useThemeStyles from '@hooks/useThemeStyles';
+
+type WorkspaceTravelNoVBAViewProps = {
+    policyID: string;
```
Comment (VickyStash) [inline src/pages/workspace/travel/WorkspaceTravelNoVBAView.tsx:11]:
```suggestion
    /** The policy ID currently being configured */
    policyID: string;
```

File: src/pages/workspace/travel/WorkspaceTravelPage.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import type {StackScreenProps} from '@react-navigation/stack';
+import React from 'react';
+import useLocalize from '@hooks/useLocalize';
+import type {SettingsNavigatorParamList} from '@libs/Navigation/types';
+import WorkspacePageWithSections from '@pages/workspace/WorkspacePageWithSections';
+import CONST from '@src/CONST';
+import WorkspaceTravelNoVBAView from './WorkspaceTravelNoVBAView';
+import WorkspaceTravelVBAView from './WorkspaceTravelVBAView';
+
+function WorkspaceTravelPage({route}: StackScreenProps<SettingsNavigatorParamList, 'Workspace_Travel'>) {
```
Comment (VickyStash) [inline src/pages/workspace/travel/WorkspaceTravelPage.tsx:10]:
```suggestion
type WorkspaceTravelPageProps = StackScreenProps<SettingsNavigatorParamList, typeof SCREENS.WORKSPACE.TRAVEL>

function WorkspaceTravelPage({route}: WorkspaceTravelPageProps) {
```

File: src/pages/workspace/withPolicy.tsx [hunk 3]
```diff
@@ -97,11 +91,11 @@ const policyDefaultProps: WithPolicyOnyxProps = {
 /*
  * HOC for connecting a policy in Onyx corresponding to the policyID in route params
  */
-export default function <TProps extends WithPolicyProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
+export default function <TRef, TProps extends WithPolicyProps>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
```
Comment (blazejkustra) [inline src/pages/workspace/withPolicy.tsx:94]: Why was TRef moved?

File: src/pages/workspace/withPolicy.tsx [hunk 4]
```diff
@@ -1,23 +1,17 @@
...
 import CONST from '@src/CONST';
 import ONYXKEYS from '@src/ONYXKEYS';
-import type SCREENS from '@src/SCREENS';
 import type * as OnyxTypes from '@src/types/onyx';
 
-type WorkspaceParamList = BottomTabNavigatorParamList & CentralPaneNavigatorParamList & SettingsNavigatorParamList;
-type PolicyRoute = RouteProp<WorkspaceParamList, ValueOf<typeof SCREENS.WORKSPACE>>;
-
-function getPolicyIDFromRoute(route: PolicyRoute): string {
+function getPolicyIDFromRoute(route?: RouteProp<Record<string, {policyID?: string}>, string>): string {
```
Comment (blazejkustra) [inline src/pages/workspace/withPolicy.tsx:14]: I think it was clearer before, why was this changed?

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/pages/workspace/withPolicy.tsx (hunk 3 above)
Comment (pac-guerreiro) [reply src/pages/workspace/withPolicy.tsx:94]: @blazejkustra I was trying something that required me to move it but then I reverted it back, so I can move it back!

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 5]
```diff
@@ -41,9 +40,6 @@ type WorkspacePageWithSectionsProps = WithPolicyAndFullscreenLoadingProps &
         /** The text to display in the header */
         headerText: string;
 
-        /** The route object passed to this page from the navigator */
-        route: PolicyRoute;
```
Comment (kubabutkiewicz) [inline src/pages/workspace/WorkspacePageWithSections.tsx:45]: why it was removed?

File: src/pages/workspace/travel/WorkspaceTravelPage.tsx [hunk 6]
```diff
@@ -1,40 +1,32 @@
...
             shouldUseScrollView
-            headerText={props.translate('workspace.common.travel')}
-            route={props.route}
+            headerText={translate('workspace.common.travel')}
+            route={route}
             guidesCallTaskID={CONST.GUIDES_CALL_TASK_IDS.WORKSPACE_TRAVEL}
             shouldShowOfflineIndicatorInWideScreen
         >
-            {(hasVBA, policyID) => (
+            {(hasVBA = false, policyID = '') => (
```
Comment (kubabutkiewicz) [inline src/pages/workspace/travel/WorkspaceTravelPage.tsx:29]: why this default parameters was added?

File: src/pages/workspace/travel/WorkspaceTravelVBAView.tsx [hunk 7]
```diff
@@ -4,61 +4,58 @@ import * as Expensicons from '@components/Icon/Expensicons';
...
                     icon: Expensicons.Info,
                     shouldShowRightIcon: true,
                     iconRight: Expensicons.NewWindow,
-                    wrapperStyle: [styles.cardMenuItem],
+                    wrapperStyle: styles.cardMenuItem,
                     link: CONST.CONCIERGE_TRAVEL_URL,
                 },
             ]}
         >
             <View style={[styles.mv3]}>
```
Comment (kubabutkiewicz) [inline src/pages/workspace/travel/WorkspaceTravelVBAView.tsx:52]:
```suggestion
            <View style={styles.mv3}>
```

File: src/pages/workspace/withPolicy.tsx [hunk 8]
```diff
@@ -99,9 +93,9 @@ const policyDefaultProps: WithPolicyOnyxProps = {
  */
 export default function <TProps extends WithPolicyProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>): React.ComponentType<Omit<TProps, keyof WithPolicyOnyxProps>> {
     function WithPolicy(props: TProps, ref: ForwardedRef<TRef>) {
-        const routes = useNavigationState((state) => state.routes || []);
-        const currentRoute = routes?.at(-1);
-        const policyID = getPolicyIDFromRoute(currentRoute as PolicyRoute);
+        const routes = useNavigationState((state: NavigationState<Record<string, {policyID?: string}>>) => state.routes);
```
Comment (kubabutkiewicz) [inline src/pages/workspace/withPolicy.tsx:96]: and also this `Record<string, {policyID?: string}>` can be move to seperate type

File: src/pages/workspace/WorkspacePageWithSections.tsx (hunk 5 above)
Comment (pac-guerreiro) [reply src/pages/workspace/WorkspacePageWithSections.tsx:45]: @kubabutkiewicz Because it's inherited from `WithPolicyAndFullscreenLoadingProps`

File: src/pages/workspace/travel/WorkspaceTravelPage.tsx (hunk 6 above)
Comment (pac-guerreiro) [reply src/pages/workspace/travel/WorkspaceTravelPage.tsx:29]: @kubabutkiewicz I think that previously these parameters could be undefined so I added the default values, but it's no longer necessary so I'll just remove it

**Comment by pac-guerreiro** [comment]
@blazejkustra @kubabutkiewicz All feedback addressed, thanks 😄

**Review by blazejkustra — APPROVED** [review]

**Review by kubabutkiewicz — APPROVED** [review]

**Review by neil-marcellini — APPROVED** [review]
Looks good. I'm going to assign a C+ to do the testing.

**Review by hoangzinh — APPROVED** [review]
LGTM
